### PR TITLE
Prefetch page index (#4090)

### LINKED ISSUE #4090: Preload page index for async ParquetObjectReader
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently the [`ParquetMetaData`](https://docs.rs/parquet/latest/parquet/file/metadata/struct.ParquetMetaData.html) object has optional fields for the column & offset indexes which are unpopulated at first. When the `ArrowReaderBuilder` is created using `ArrowReaderOptions::with_page_index(true)` it loads the page index at query time. However, this is potentially suboptimal as it incurs additional latency making an extra request (typically to object storage which is high-latency) for each query. 

**Describe the solution you'd like**
A new method for the `ParquetObjectReader` that toggles loading the page index at construction time, something like this:
```rust
impl ParquetObjectReader {
    pub fn preload_page_index(self, should_preload: bool) -> Self {
        self.preload_page_index = true
    }
}
```

which would trigger conditional logic in the `get_metadata` function to return metadata with the page index already loaded. 

**Describe alternatives you've considered**
A public async API for deserializing the column & offset index, similar to [`index_reader`](https://docs.rs/parquet/latest/parquet/file/page_index/index_reader/index.html) but with async support and integrated with `AsyncFileReader` to enable coalescing of multiple fetches.

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4090

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

This adds `Send` constraints to `fetch_parquet_metadata`, this is unlikely to trip people up in practice

### CODE REVIEW

File: parquet/src/arrow/async_reader/metadata.rs [hunk 1]
```diff
@@ -15,95 +15,252 @@
...
+}
+
+impl<'a, T: AsyncFileReader> MetadataFetch for &'a mut T {
+    fn fetch(&mut self, range: Range<usize>) -> BoxFuture<'_, Result<Bytes>> {
+        self.get_bytes(range)
     }
+}
+
+/// An asynchronous interface to load [`ParquetMetaData`] from an async source
+pub(crate) struct MetadataLoader<F> {
```
Comment (tustvold) [inline parquet/src/arrow/async_reader/metadata.rs:45]: I opted to keep this crate private for now as I'm not totally happy with the interface

**Review by alamb — APPROVED** [review]
Thank you @tustvold  -- I went through this PR carefully and it looks really nice 👌 

I wonder if the same basic pattern could be applied to the Bloom filters as well, or if they suffer from the issue that they don't actually appear in the footer 🤔 

cc @thinkharderdev and @Ted-Jiang

**Comment by Ted-Jiang** [comment]
> Thank you @tustvold -- I went through this PR carefully and it looks really nice 👌
> 
> I wonder if the same basic pattern could be applied to the Bloom filters as well, or if they suffer from the issue that they don't actually appear in the footer 🤔
> 
> cc @thinkharderdev and @Ted-Jiang

Thanks for ping me, i will review this carefully today.

File: parquet/src/arrow/async_reader/metadata.rs [hunk 2]
```diff
@@ -15,95 +15,250 @@
...
+
+        // Did not fetch the entire file metadata in the initial read, need to make a second request
+        let (metadata, remainder) = if length > suffix_len - 8 {
+            let metadata_start = file_size - length - 8;
+            let remaining_metadata = fetch.fetch(metadata_start..footer_start).await?;
+
+            let reader = remaining_metadata.as_ref().chain(&suffix[..suffix_len - 8]);
+            (read_metadata(reader)?, None)
+        } else {
+            let metadata_start = file_size - length - 8 - footer_start;
```
Comment (Ted-Jiang) [inline parquet/src/arrow/async_reader/metadata.rs:114]: why need minus  `footer_start ` 🤔 . Not consist with datafusion https://github.com/apache/arrow-datafusion/blob/6e819d6c2b9280198c67fa16df3e54c79ce46ca2/datafusion/core/src/datasource/file_format/parquet.rs#L420

Comment (Ted-Jiang) [reply parquet/src/arrow/async_reader/metadata.rs:114]: Oh, It is total different logic 🤣  my mistake!

**Review by Ted-Jiang — APPROVED** [review]
👍

**Comment by tustvold** [comment]
> I wonder if the same basic pattern could be applied to the Bloom filters as well

Yes, the design of this was with them in mind. Bloom filters can be stored at the end of the file, which would allow prefetching to help, I'm not sure the writer currently does this though

**Pull request merged**